bootloaders/uart/uart_boot.c

```diff
@@ -0,0 +1,218 @@
+        // Detect hangs and reset the chip
+        printf("Clear has hung - resetting\n");
+        reset_chip();
+        return;
+    }
+    in = uart_getc(UART_ID);
+    printf("%c\n", in);
+    if (in != 'c') {
+        printf("Incorrect response - resetting\n");
+        reset_chip();
+        return;
+    }
+
+    printf("Verifying binary\n");
+    time_start = time_us_32();
+    current_addr = start_addr;
+    while (current_addr < end_addr) {
+        uart_putc_raw(UART_ID, 'r');
+        char *buf = (char*)current_addr;
+        if (!uart_is_readable_within_us(UART_ID, 500)) {
+            // Detect hangs and reset the chip
+            printf("Verify has hung - resetting\n");
+            reset_chip();
+            return;
+        }
+        int i = 0;
+        while (uart_is_readable_within_us(UART_ID, 10) && i < 32) {
+            in = uart_getc(UART_ID);
+            if (in != buf[i]) {
+                printf("Verify has incorrect data at 0x%08x - resetting\n", current_addr - start_addr + SRAM_BASE);
+            }
+            i++;
+        }
+        if (i != 32) {
+            printf("Verify has incorrect data size - resetting\n");
+        }
+        in = uart_getc(UART_ID);
+        if (in != 'r') {
+            printf("Incorrect response - resetting\n");
+            reset_chip();
+            return;
+        }
+        current_addr += 32;
+    }
+
+    time_end = time_us_32();
+    printf("Verify took %dus\n", time_end - time_start);
+    printf("Verify complete - executing\n");
+
+    uart_putc_raw(UART_ID, 'x');
+    if (!uart_is_readable_within_us(UART_ID, 500)) {
+        // Detect hangs and reset the chip
+        printf("Execute has hung - resetting\n");
+        reset_chip();
+        return;
+    }
+    in = uart_getc(UART_ID);
+    printf("%c\n", in);
+    if (in != 'x') {
+        printf("Incorrect response - resetting\n");
+        reset_chip();
+        return;
+    }
+}
+
+
+int main()
+{
+    stdio_init_all();
+
+    // Set up our UART for booting the other device
+    uart_init(UART_ID, 1000000);
+    gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+    gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+
+    // Set up run pin
+    gpio_init(RUN_PIN);
+    gpio_set_dir(RUN_PIN, GPIO_OUT);
+
+    // Reset chip
+    reset_chip();
+
+
+    while (true) {
+        char splash[] = "RP2350";
+        char hello[] = "Hello";
+        char buf[500] = {0};
+        int i = 0;
+        while (uart_is_readable(UART_ID) && i < sizeof(buf)) {
+            char in = uart_getc(UART_ID);
+            printf("%c", in);
+            buf[i] = in;
+            i++;
+        }
+        if (i > 0) {
+            printf(" ...Read done\n");
+        }
+        char *ptr = memchr(buf, splash[0], sizeof(buf));
+        if (ptr && strncmp(ptr, splash, sizeof(splash) - 1) == 0) {
+            printf("Splash found\n");
+            uart_boot();
+        } else {
+            ptr = memchr(buf, hello[0], sizeof(buf));
```

Comment:
Hah, I guess it's important that `hello[0]` doesn't appear anywhere in the `splash` string :wink: 

Is there any reason not to have this check for the full "Hello, world\n" that `uart_binary` transmits?